### PR TITLE
Auto-queue videos from library

### PR DESCRIPTION
**Pull request by bcedu**
# Binge mode

When a file is played from the library, all next episodes are added to the queue. By this, users can star to watch a show without the need of select the next episode every time an episode finish.

imgl1

imgl2

## Other changes

- When a file is played from the library, the playlist is cleared before adding all the next episodes of the selected file.

## Related

Old PR https://github.com/elementary/videos/pull/122

### CODE REVIEW

File: src/Widgets/Library/EpisodesPage.vala [hunk 1]
```diff
@@ -73,10 +74,19 @@ namespace Audience {
             view_episodes.forall ((item) => {
                 item.dispose ();
             });
+            shown_episodes = new Gee.ArrayList<Audience.Objects.Video> ();
             foreach (Audience.Objects.Video episode in episodes) {
                 view_episodes.add (new Audience.LibraryItem (episode, LibraryItemStyle.ROW));
+                shown_episodes.add(episode);
```
Comment (ryonakano) [inline src/Widgets/Library/EpisodesPage.vala]:
```suggestion
                shown_episodes.add (episode);
```

File: src/Widgets/Library/EpisodesPage.vala [hunk 2]
```diff
@@ -73,10 +74,19 @@ namespace Audience {
...
             }
-
+            shown_episodes.sort((a, b) => {
```
Comment (ryonakano) [inline src/Widgets/Library/EpisodesPage.vala]:
```suggestion
            shown_episodes.sort ((a, b) => {
```

File: src/Widgets/Library/EpisodesPage.vala [hunk 3]
```diff
@@ -95,9 +105,15 @@ namespace Audience {
             if (video.video_file.query_exists ()) {
                 string uri = video.video_file.get_uri ();
                 bool from_beginning = uri != settings.get_string ("current-video");
-                var window = App.get_instance ().mainwindow;
+                var window = App.get_instance ().mainwindow;// Clean playlist
+                window.clear_playlist();
```
Comment (ryonakano) [inline src/Widgets/Library/EpisodesPage.vala]:
```suggestion
                window.clear_playlist ();
```

File: src/Widgets/Library/EpisodesPage.vala [hunk 4]
```diff
@@ -95,9 +105,15 @@ namespace Audience {
...
                 window.add_to_playlist (uri, false);
                 window.play_file (uri, Window.NavigationPage.EPISODES, from_beginning);
+                // Add next from the current view to the queque
+                int played_index = shown_episodes.index_of(video);
```
Comment (ryonakano) [inline src/Widgets/Library/EpisodesPage.vala]:
```suggestion
                int played_index = shown_episodes.index_of (video);
```

File: src/Widgets/Library/EpisodesPage.vala [hunk 5]
```diff
@@ -95,9 +105,15 @@ namespace Audience {
...
+                foreach (Audience.Objects.Video episode in shown_episodes.slice(played_index, shown_episodes.size)) {
```
Comment (ryonakano) [inline src/Widgets/Library/EpisodesPage.vala]:
```suggestion
                foreach (Audience.Objects.Video episode in shown_episodes.slice (played_index, shown_episodes.size)) {
```

File: src/Widgets/Library/EpisodesPage.vala [hunk 6]
```diff
@@ -95,9 +105,15 @@ namespace Audience {
...
+                    window.append_to_playlist(episode.video_file);
```
Comment (ryonakano) [inline src/Widgets/Library/EpisodesPage.vala]:
```suggestion
                    window.append_to_playlist (episode.video_file);
```

File: src/Window.vala [hunk 7]
```diff
@@ -490,6 +490,14 @@ public class Audience.Window : Gtk.Window {
         welcome_page.refresh ();
     }
 
+    public void clear_playlist() {
```
Comment (ryonakano) [inline src/Window.vala]:
```suggestion
    public void clear_playlist () {
```

File: src/Window.vala [hunk 8]
```diff
@@ -490,6 +490,14 @@ public class Audience.Window : Gtk.Window {
...
+        player_page.get_playlist_widget().clear_items();
```
Comment (ryonakano) [inline src/Window.vala]:
```suggestion
        player_page.get_playlist_widget ().clear_items ();
```

File: src/Window.vala [hunk 9]
```diff
@@ -490,6 +490,14 @@ public class Audience.Window : Gtk.Window {
...
+    }
+
+    public void append_to_playlist(File file) {
```
Comment (ryonakano) [inline src/Window.vala]:
```suggestion
    public void append_to_playlist (File file) {
```

File: src/Window.vala [hunk 10]
```diff
@@ -490,6 +490,14 @@ public class Audience.Window : Gtk.Window {
...
+        player_page.append_to_playlist(file);
```
Comment (ryonakano) [inline src/Window.vala]:
```suggestion
        player_page.append_to_playlist (file);
```

**Comment by arshubham** [comment]
@bcedu Nice start. I think it would be worth adding a mode switch which is shown on header bar with an associated gsettings (off by default), for if a user want to play just a single video from the library.

**Comment by bcedu** [comment]
@arshubham Nice. Which icon should I put next to the switch button?

**Comment by arshubham** [comment]
@bcedu You can just add any icons for now. Maybe @elementary/ux  will suggest some icons here. Couldn't find any relevant icons here. For now maybe you can use `camera-video-symbolic` for binge more and `media-video-symbolic` for normal

**Comment by cassidyjames** [comment]
@arshubham never choose icons for how they work! They should be as semantically-named as possible. In this case, I could see us creating bespoke `media-autoplay-on` and `media-autoplay-off` icons, but for now we can probably use `media-playback-pause-symbolic` for off (since it'll pause after playing one video), and `media-playback-skip-forward` for autoplay since it effectively skips to the next video when it's on.

**Comment by cassidyjames** [comment]
I'm also not sure if the headerbar makes sense here, or if it should just be in the queue/playlist popover.

**Comment by cassidyjames** [comment]
Alternatively we could just add an autoplay togglebutton next to the repeat button in the playlist popover.

**Comment by bcedu** [comment]
I've made the button in the header. I think it's the good place for it, by putting it on the header bar the use can select if he wants to queue all the files while is choosing the episodie from the library.

If we put it in the popover the user will haveto change it when he is already playing the file.

This is how it looks:

- In the welcome screen the switch is not shown:

c1

- When the user goes to library the switch is shown:

c3

- It is also shown while playing a file:

c5

- If the user goes baxk to the welcome screen the switch is hidden

**Comment by bcedu** [comment]
Obviously if you decide to place the button in the popover I will move it to the popover

File: src/Window.vala [hunk 11]
```diff
@@ -77,6 +78,12 @@ public class Audience.Window : Gtk.Window {
 
         header.pack_end (search_entry);
 
+        autoqueque_next = new Granite.ModeSwitch.from_icon_name ("media-playback-pause-symbolic", "media-playback-skip-forward");
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling and icon name

```suggestion
        autoqueue_next = new Granite.ModeSwitch.from_icon_name ("media-playback-pause-symbolic", "media-skip-forward-symbolic");
```

File: src/Window.vala [hunk 12]
```diff
@@ -33,6 +33,7 @@ public class Audience.Window : Gtk.Window {
     private NavigationButton navigation_button;
     private ZeitgeistManager zeitgeist_manager;
     private Gtk.SearchEntry search_entry;
+    private Granite.ModeSwitch autoqueque_next;
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling

```suggestion
    private Granite.ModeSwitch autoqueue_next;
```

File: src/Window.vala [hunk 13]
```diff
@@ -77,6 +78,12 @@ public class Audience.Window : Gtk.Window {
 
         header.pack_end (search_entry);
 
+        autoqueque_next = new Granite.ModeSwitch.from_icon_name ("media-playback-pause-symbolic", "media-playback-skip-forward");
+        autoqueque_next.primary_icon_tooltip_text = _("Standard mode: play single video");
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling and string change

```suggestion
        autoqueue_next.primary_icon_tooltip_text = _("Play one video");
```

File: src/Window.vala [hunk 14]
```diff
@@ -77,6 +78,12 @@ public class Audience.Window : Gtk.Window {
...
+        autoqueque_next.secondary_icon_tooltip_text = _("Binge mode: queque next files from library");
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling and string change

```suggestion
        autoqueue_next.secondary_icon_tooltip_text = _("Automatically play next videos");
```

File: src/Window.vala [hunk 15]
```diff
@@ -77,6 +78,12 @@ public class Audience.Window : Gtk.Window {
...
+        autoqueque_next.valign = Gtk.Align.CENTER;
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling

```suggestion
        autoqueue_next.valign = Gtk.Align.CENTER;
```

File: src/Window.vala [hunk 16]
```diff
@@ -77,6 +78,12 @@ public class Audience.Window : Gtk.Window {
...
+        header.pack_end (autoqueque_next);
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling

```suggestion
        header.pack_end (autoqueue_next);
```

File: src/Window.vala [hunk 17]
```diff
@@ -110,6 +117,7 @@ public class Audience.Window : Gtk.Window {
                 main_stack.set_visible_child (episodes_page);
                 this.title = item.get_title ();
                 search_entry.text = "";
+                autoqueque_next.visible = true;
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling

```suggestion
                autoqueue_next.visible = true;
```

File: src/Window.vala [hunk 18]
```diff
@@ -169,6 +177,7 @@ public class Audience.Window : Gtk.Window {
 
         navigation_button.hide ();
         search_entry.visible = false;
+        autoqueque_next.visible = false;
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling

```suggestion
        autoqueue_next.visible = false;
```

File: src/Window.vala [hunk 19]
```diff
@@ -474,6 +483,7 @@ public class Audience.Window : Gtk.Window {
                 break;
             case NavigationPage.EPISODES:
                 navigation_button.label = _(NAVIGATION_BUTTON_EPISODES);
+                autoqueque_next.visible = true;
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling

```suggestion
                autoqueue_next.visible = true;
```

File: src/Window.vala [hunk 20]
```diff
@@ -505,13 +523,16 @@ public class Audience.Window : Gtk.Window {
         if (navigation_button.label == _(NAVIGATION_BUTTON_LIBRARY)) {
             navigation_button.label = _(NAVIGATION_BUTTON_WELCOMESCREEN);
             main_stack.set_visible_child_full ("library", Gtk.StackTransitionType.SLIDE_RIGHT);
+            autoqueque_next.visible = false;
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling

```suggestion
            autoqueue_next.visible = false;
```

File: src/Window.vala [hunk 21]
```diff
@@ -505,13 +523,16 @@ public class Audience.Window : Gtk.Window {
...
         } else if (navigation_button.label == _(NAVIGATION_BUTTON_EPISODES)) {
             navigation_button.label = _(NAVIGATION_BUTTON_LIBRARY);
             main_stack.set_visible_child_full ("episodes", Gtk.StackTransitionType.SLIDE_RIGHT);
+            autoqueque_next.visible = true;
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling

```suggestion
            autoqueue_next.visible = true;
```

File: src/Window.vala [hunk 22]
```diff
@@ -505,13 +523,16 @@ public class Audience.Window : Gtk.Window {
...
         } else {
             navigation_button.hide ();
             main_stack.set_visible_child (welcome_page);
             search_entry.visible = false;
+            autoqueque_next.visible = false;
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling

```suggestion
            autoqueue_next.visible = false;
```

File: src/Window.vala [hunk 23]
```diff
@@ -548,4 +569,8 @@ public class Audience.Window : Gtk.Window {
     public void show_mouse_cursor () {
         get_window ().set_cursor (null);
     }
+
+    public bool autoqueque_next_active () {
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling

```suggestion
    public bool autoqueue_next_active () {
```

File: src/Window.vala [hunk 24]
```diff
@@ -548,4 +569,8 @@ public class Audience.Window : Gtk.Window {
...
+        return this.autoqueque_next.active;
```
Comment (cassidyjames) [inline src/Window.vala]: Spelling

```suggestion
        return this.autoqueue_next.active;
```

File: src/Widgets/Library/EpisodesPage.vala [hunk 25]
```diff
@@ -95,9 +105,17 @@ namespace Audience {
             if (video.video_file.query_exists ()) {
                 string uri = video.video_file.get_uri ();
                 bool from_beginning = uri != settings.get_string ("current-video");
-                var window = App.get_instance ().mainwindow;
+                var window = App.get_instance ().mainwindow;// Clean playlist
+                window.clear_playlist ();
                 window.add_to_playlist (uri, false);
                 window.play_file (uri, Window.NavigationPage.EPISODES, from_beginning);
+                if (window.autoqueque_next_active ()) {
```
Comment (cassidyjames) [inline src/Widgets/Library/EpisodesPage.vala]: Spelling

```suggestion
                if (window.autoqueue_next_active ()) {
```

File: src/Widgets/Library/EpisodesPage.vala [hunk 26]
```diff
@@ -95,9 +105,17 @@ namespace Audience {
...
+                    // Add next from the current view to the queque
```
Comment (cassidyjames) [inline src/Widgets/Library/EpisodesPage.vala]: Spelling

```suggestion
                    // Add next from the current view to the queue
```

File: src/Window.vala [hunk 27]
```diff
@@ -77,6 +78,18 @@ public class Audience.Window : Gtk.Window {
 
         header.pack_end (search_entry);
 
+        autoqueue_next = new Granite.ModeSwitch.from_icon_name ("media-playback-pause-symbolic", "media-playback-skip-forward");
+        autoqueue_next.primary_icon_tooltip_text = _("Play one video");
+        autoqueue_next.secondary_icon_tooltip_text = _("Automatically play next videos");
+        autoqueue_next.valign = Gtk.Align.CENTER;
+        autoqueue_next.active = settings.get_boolean ("autoqueue-next");
+        this.delete_event.connect(() => {
```
Comment (cassidyjames) [inline src/Window.vala]: Should probably just bind to the GSetting instead

File: data/io.elementary.videos.gschema.xml [hunk 28]
```diff
@@ -59,5 +59,10 @@
             <summary>Poster file names</summary>
             <description></description>
         </key>
+        <key name="autoqueue-next" type="b">
+            <default>true</default>
+            <summary>Don't queque next videos when playing a file from library.</summary>
```
Comment (cassidyjames) [inline data/io.elementary.videos.gschema.xml]: The setting is set to "true", but the summary says "don't", and the description says it's deactivated. But the code doesn't appear to reflect that. This is super confusing.

```suggestion
            <summary>Automatically play next videos</summary>
```

File: data/io.elementary.videos.gschema.xml [hunk 29]
```diff
@@ -59,5 +59,10 @@
...
+            <description>This option allows user to activate the autoqueque videos when playing a video from library. Deactivated by default.</description>
```
Comment (cassidyjames) [inline data/io.elementary.videos.gschema.xml]:
```suggestion
            <description>Automatically queue up next videos by adding them to the playlist when playing from the library.</description>
```

**Pull request merged**